README.md
# Anymate

Api for users and uploaded 3D models V1

## Description

list of endpoints to retrieve, delete and push user data to the db, users can post a model, edit and delete it from the models db

## Getting Started

Make an .env file with a POSTGRES_PASSWORD &  POSTGRES_USER & POSTGRES_DB
to execute  the program use this command: docker-compose up --build

### Dependencies

* dotenv
* express
* jest
* knex
* nodemon
* pg
* supertest
* uuid

endpoints: <br />
GET /users <br />
``` This returns all users in the DB``` <br />
GET /user/:uuid <br />
``` This returns the user based on uuid:
  http://localhost:3000/user/a73ffd78-44a0-11ec-81d3-2497ac133203
  returns:
  [
    {
        "uuid": "a73ffd78-44a0-11ec-81d3-2497ac133203",
        "name": "name",
        "age": "30"
    }
]
``` 
POST /user <br />
``` Saves user to db (send json object)

{
    "uuid": "a73ffd78-99a0-11ec-81d3-2497ac133203", 
    "name": "Omar", 
    "age":"25"
}
``` 
PATCH /user/:uuid <br />
```edit user data based on uuid:
http://localhost:3000/user/a73ffd78-44a0-11ec-81d3-2497ac133203
{
    "name": "a",
    "age": "13"
}

```
DELETE /user/:uuid <br />
``` Delete user based on uuid
http://localhost:3000/user/a73ffd78-44a0-11ec-81d3-2497ac133203
returns deleted user as confirmation: 
[
    {
        "uuid": "a73ffd78-44a0-11ec-81d3-2497ac133203",
        "name": "a",
        "age": "13"
    }
]
```

GET /models <br />
``` This returns all models in the DB``` <br />
GET /model/:uuid <br />
``` This returns the model based on uuid:
  http://localhost:3000/model/a63ffd78-44a0-11ec-81d3-3586ac130003
  returns:
 [
    {
        "uuid": "a63ffd78-44a0-11ec-81d3-3586ac130003",
        "title": "Running man",
        "CreatedBy": "Omar"
    }
]
``` 
POST /model <br />
``` Saves model to db (send json object)

{
    "uuid": "a63ffd78-44a0-11ec-81d3-3586ac130003",
    "title": "Running man",
    "CreatedBy": "Omar"
}
``` 
PATCH /model/:uuid <br />
```edit model data based on uuid:
http://localhost:3000/model/a63ffd78-44a0-11ec-81d3-3586ac130003
{
    "title": "this is a new title"
}

```
DELETE /model/:uuid <br />
``` Delete model based on uuid
http://localhost:3000/model/a63ffd78-44a0-11ec-81d3-3586ac130003
returns deleted model as confirmation: 
[
    {
        "uuid": "a63ffd78-44a0-11ec-81d3-3586ac130003",
        "title": "this is a new title",
        "CreatedBy": "Omar"
    }
]
```

## Authors

Omar Aghallaj


CHANGELOG.md
# Changelog

* Ver 1.0.0
Added document files & structure

* Ver 1.1.0
Added a basic express server with endpoints

* Ver 1.2.0
Updated endpoints & added UUID

* Ver 1.2.1
added packages & adjumestments on documents (dockerfile, readme )

* Ver 1.3.1
Integrated first unit- & integration tests

* Ver 1.3.2
Added docs (knexfile & migrations), adjusted readme

* Ver 1.3.3
Fixed db-data leak in dockerfile & path

* Ver 1.3.4
Env file variable edited & filename

* Ver 1.3.5
workdir path fixed in dockerfile

* Ver 1.4.5
createTable helper file created & adjusted knex config

* Ver 1.4.6
docker-compose error connection fixed

* Ver 1.4.7
Removed "postgres" dependency

* Ver 1.5.7
implemented complete integration tests on endpoints

* Ver 1.6.7
Added persistence to db data

* Ver 1.7.7
Finished all endpoints

* Ver 1.7.8
small logging fix & conventions

* Ver 1.8.8
Completed integration test for user endpoints

* Ver 1.9.8
Completed integration test for model endpoints

* Ver 1.10.8
Completed end to end tests

* Ver 1.10.9
Edited Readme file to be up to date


api/src/__tests__/unit.test.js
const key = require('../uuid');

describe('uuid test', () =>{
    test('check if the uuid is valid', ()=>{
        expect(key.generateUUID()).toMatch(/^[0-9A-F]{8}-[0-9A-F]{4}-[1][0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i);
    });
    test('check if uuid is defined', ()=>{
        expect(key.generateUUID()).toBeDefined();
    });
    test('Check if uuid can not be equal', ()=>{
        let uniqueId = key.generateUUID();
        expect(key.generateUUID()).not.toEqual(uniqueId);
    })
});


api/src/__tests__/endToEnd.test.js
const supertest = require('supertest');
const app = require('../index')
const request = supertest(app);
const key = require('../uuid');
jest.useFakeTimers()

describe('end to end test', ()=>{
    const id = key.generateUUID();
    const idN = key.generateUUID();

    test('create user', (done)=>{
        request.post('/user')
        .send({uuid: id, name: "omarTest", age: "25"})
        .then((response)=>{
            expect(response.status).toBe(200);
            expect(typeof response.body[0].name).toBe("string")
            expect(typeof response.body[0].age).toBe("string")
            done();
        })
        .catch((e)=>{
            done(e)
        })
    })

    test('create user with missing data', (done)=>{
        request.post('/user')
        .send({uuid: id,name: "", age: "25"})
        .then((response)=>{
            expect(response.status).toBe(400);
            expect(response.body[0].name).toBe(undefined);
            done();
        }) 
        .catch((e)=>{
            done(e)
        })
    })

    test('create user with duplicate uuid', (done)=>{
        request.post('/user')
        .send({uuid: id, name: "Omar", age: "20"})
        .then((response)=>{
            console.log(response.text);
            expect(response.status).toBe(400);
            done();
        }) 
        .catch((e)=>{
            done(e)
        })
    });
    test('logged in user', async()=>{
        const data = await request.get(`/user/${id}`);
        expect(data.body[0].name).not.toBeNaN();
        expect(data.body[0].name && data.body[0].age).toBeTruthy();
    })
    test('Edit user name', (done)=>{
        request.patch(`/user/${id}`)
        .send({uuid: id, name: "sasha"})
        .then((response)=>{
            expect(response.status).toBe(200);
            done();
        }) 
        .catch((e)=>{
            done(e)
        })
    })

    test('edited user fetch', async()=>{
        const data = await request.get(`/user/${id}`);
        expect(data.body[0].name).toBe("sasha");
    })

    test('post a model', (done)=>{
        request.post('/model')
        .send({uuid: idN, title: "running man", CreatedBy:"Omar"})
        .then((response)=>{
            expect(response.status).toBe(200);
            expect(typeof response.body[0].title).toBe("string")
            expect(typeof response.body[0].CreatedBy).toBe("string")
            expect(response.body[0].title).not.toBeNull()
            expect(response.body[0].CreatedBy).not.toBeNull()
            done();
        })
        .catch((e)=>{
            done(e)
        })
    })
    test('Edit model name', (done)=>{
        request.patch(`/model/${idN}`)
        .send({uuid: idN, title: "walking man"})
        .then((response)=>{
            expect(response.status).toBe(200);
            done();
        }) 
        .catch((e)=>{
            done(e)
        })
    })
    test('delete a model', (done)=>{
        request.delete(`/model/${idN}`)
        .then((resp)=>{
            expect(resp.status).toBe(200);
            done();
        })
        .catch((e)=>{
            done(e)
        }); 
    });
})

api/src/__tests__/integration.test.js
const supertest = require('supertest');
const app = require('../index')
const request = supertest(app);
const key = require('../uuid');
jest.useFakeTimers()

describe ("save user data to db, fetch all users, fetch, edit and delete the user", ()=>{
    const idP = key.generateUUID();
    test('returns status code 200 if user is passed', (done) =>{
        request.post('/user')
        .send({uuid: idP, name: "omar", age:"25"})
        .then((response)=>{
            expect(response.status).toBe(200);
            done();
        })
        .catch((e)=>{
            done(e)
        })
    });

    test('GET /users returns users from db & 200', async ()=>{
        const resp = await request.get('/users');
        expect(resp.statusCode).toBe(200);
        expect(typeof resp.body).toBe("object");
    });

    test('GET /user/:uuid returns user based on uuid', async ()=>{
        const resp = await request.get(`/user/${idP}`);
        expect(resp.statusCode).toBe(200);
        expect(typeof resp.body).toBe("object");
    });

    test('patch /user/:uuid returns modified user data based on uuid', (done)=>{
        request.patch(`/user/${idP}`)
        .send({uuid: idP, name: "Omar edited"})
        .then((resp)=>{
            expect(resp.status).toBe(200);
            done();
        }).catch((resp)=>{
            
            done()
        })

    })
    test('DELETE /user/:uuid', (done)=>{
        request.delete(`/user/${idP}`)
        .then((resp)=>{
            expect(resp.status).toBe(200);
            done();
        })
        .catch((e)=>{
            
            done(e)
        })
    });
});

jest.useFakeTimers()
describe('insert model data, get all models, get, edit and delete model', ()=>{
    const idM = key.generateUUID();
    test('returns status code 200 if model is passed', (done) =>{
        request.post('/model')
        .send({uuid: idM, title: "running man", CreatedBy:"Omar"})
        .then((response)=>{
            expect(response.status).toBe(200);
            done();
        })
        .catch((e)=>{
            done(e)
        })
    });

    test('GET /models returns models from db & 200', async ()=>{
        const resp = await request.get('/models');
        expect(resp.statusCode).toBe(200);
        expect(typeof resp.body).toBe("object");
    });

    test('GET /model/:uuid returns model based on uuid', async ()=>{
        const resp = await request.get(`/model/${idM}`);
        expect(resp.statusCode).toBe(200);
        expect(typeof resp.body).toBe("object");
    });

    test('patch /model/:uuid returns modified model data based on uuid', (done)=>{
        request.patch(`/model/${idM}`)
        .send({uuid: idM, title: "edited title"})
        .then((resp)=>{
            expect(resp.status).toBe(200);
            done();
        }).catch((e)=>{
            done(e)
        })

    })
    test('DELETE /model/:uuid', (done)=>{
        request.delete(`/model/${idM}`)
        .then((resp)=>{
            expect(resp.status).toBe(200);
            done();
        })
        .catch((e)=>{
            done(e)
        })
    });
});

api/src/Helpers/dbHelper.js
/** Creates a table for users
 * @param
 * @returns a table user with the neccessary rows
 */

 async function createTable(db) {
    await db.schema.hasTable('users').then(async (exists) => {
      if (!exists) {
          return db.schema.createTable('users', (table) => {
            table.uuid('uuid');
            table.string('name');
            table.string('age');
          }).then(async () => {console.log('created new table users');
        });
      }else{
          console.log("users table already exists");
      }
    });
  }

/** Creates a table for models
 * @param
 * @returns a table models with the neccessary rows
 */
  async function createModelsTable(db){
    await db.schema.hasTable('models').then(async (exists)=> {
      if (!exists){
        return db.schema.createTable('models', (table)=>{
          table.uuid('uuid');
          table.string('title');
          table.string('CreatedBy');
        }).then(async ()=>{
          console.log('created new table models');
        });
      }else{
        console.log("model table already exists");
      }
    });
  }
  module.exports = {createTable, createModelsTable}

api/src/index.js
const express = require("express");
const app = express();
require('dotenv').config();
const {createTable, createModelsTable} = require("./Helpers/dbHelper");

const db = require('knex')({
  client: 'pg',
  connection: {
    host: process.env.POSTGRES_HOST,
    port: 5432,
    user: process.env.POSTGRES_USER,
    password: process.env.POSTGRES_PASSWORD,
    database: process.env.POSTGRES_DATABASE
  }
});

app.use(express.json());
app.use(express.urlencoded({extended: true}));

/** Get root route
 * 
 * @returns a message that shows the server is up and running
 */
app.get("/", async (req, res) => {
  res.json({message: "Server depoloyed!"});
});

/** Get all users
 * 
 * @returns json object of all users
 */
app.get('/users', async (req, res) => {
  const result = await db
  .select("*")
  .from('users');
  console.log(`all users in db : ${JSON.stringify(result)}`)
  res.json(result);
});

/** Get a user based on a unique identifier
 * @params send uuid as param to fetch the user
 * @returns all user data
 */
 app.get('/user/:uuid', async (req, res) => {
  const result = await db.select(['*']).from('users').where({uuid: req.params.uuid})
  console.log(`logged in as : ${JSON.stringify(result)}`)
  res.status(200).json(result)
});

/**  Creates a user with the given data
 * @params uuid, name, age - user data to be sent
 * @returns all the inserted data
 */
app.post('/user', async (req, res) => {
  const data = await db.select("*").from("users");

  let info = [];
  for(let i = 0; i<data.length; i++){
    info.push(data[i].uuid);
  }

  if (!req.body.uuid || !req.body.name || !req.body.age) {
    res.status(400).json('incomplete data');
  }else if (info.includes(req.body.uuid)){ 
    res.status(400).json('uuid exists already');
  }else {
    const result = await db.insert([{
      uuid: req.body.uuid,
      name: req.body.name,
      age: req.body.age
    }, ]).table('users').returning('*').then((res) => {
      console.log(`Added user ${req.body.name + ", " + req.body.uuid}`);
      return res;
    });
    res.send(result);
  }
});

/**
 * Change details of the user
 * @params Send data through the body to change the details that will be modified.
 * @returns Returns status 200/OK
 */
 app.put("/user/:uuid", async (req, res) => {
   if(req.params.uuid){
     const result = await db
     .table("users")
     .where({uuid: req.params.uuid})
     .update(req.body)
     res.sendStatus(200);
   }else{
     res.sendStatus(400);
   }
  
})
/** Delete all data of a user based on a unique identifier
 * @params uuid - deleting the user after incorporating the id in the route
 * @returns the deleted user
 */
app.delete('/user/:uuid', async (req, res) => {
  if (req.params.uuid){
    const result = await db
    .table('users')
    .where({uuid: req.params.uuid})
    .del('*')
    console.log(`Deleting user: ${req.params.uuid}`)
    res.status(200).send(result);
  }else{
    res.status(400).send("id not found");
  }
});

/** Delete all users
 * 
 * @returns the deleted users
 */
app.delete('/users', async (req, res) =>{
  const result = await db
  .table('users')
  .del('*')
  .then((res) => {
    return res;
  });
  res.send(result);
})


/**
 * Get all models
 * 
 * @returns all models as json object
 */
 app.get('/models', async (req, res) => {
  const result = await db
  .select("*")
  .from('models');
  res.json(result);
})

/** Get a model based on a unique identifier
 * @params send uuid as param to fetch the model
 * @returns all model data
 */
 app.get('/model/:uuid', async (req, res) => {
  const result = await db.select(['*']).from('models').where({uuid: req.params.uuid})
  res.json(result)
});
/**
 * Uploads model by user
 * @params uuid, title, creator name
 * @returns all the inserted data
 */
app.post('/model', async (req,res) =>{
  const data = await db.select("*").from("models");

  let info = [];
  for(let i = 0; i<data.length; i++){
    info.push(data[i].uuid);
  }

  if (!req.body.uuid || !req.body.title || !req.body.CreatedBy){
    res.status(400).json("incomplete data");
  }else if (info.includes(req.body.uuid)){ 
    res.status(400).json('uuid exists already')
  }else{
    const returned = await db.insert([{
      uuid: req.body.uuid,
      title: req.body.title,
      CreatedBy: req.body.CreatedBy
    }]).table('models').returning('*').then((res) => {
      console.log(`Added model ${req.body.uuid + " name: " + req.body.title + " & Created by: " +req.body.CreatedBy}`);
      return res;
    });
    res.send(returned);
  }
})

/**
 * Change details of the model
 * @params Send data through the body to change the details that will be modified.
 * @returns Returns status 200/OK
 */
app.put("/model/:uuid", async (req, res) => {
  if(req.params.uuid){
    const result = await db
    .table("models")
    .where({uuid: req.params.uuid})
    .update(req.body)
    res.sendStatus(200);
  }else{
    res.sendStatus(400);
  }
})

/** Delete all data of a model based on a unique identifier
 * @params uuid - deleting the model after incorporating the id in the route
 * @returns the deleted model
 */
 app.delete('/model/:uuid', async (req, res) => {
  if (req.params.uuid){
    const result = await db
    .table('users')
    .where({uuid: req.params.uuid})
    .del('*')
    console.log(`Deleting user: ${req.params.uuid}`)
    res.status(200).send(result);
  }else{
    res.status(400).send("id not found");
  }
});

/** Delete all models
 * 
 * @returns the deleted model
 */
 app.delete('/models', async (req, res) =>{
  const result = await db
  .table('models')
  .del('*')
  .then((res) => {
    return res;
  });
  res.send(result);
})

createModelsTable(db);
createTable(db);
module.exports = app
